### ISSUE TITLE
Race condition in wait()

### ISSUE THREAD

**Issue by hategan** [ExaWorks/job-api-spec#129]
There is an inherent race condition in `Job.wait()`, which is triggered by the "submit, wait for queued, cancel, wait for cancelled" example. The relevant code is:

```python
...
executor.submit(job)
...
job.wait([JobState.QUEUED])
...
```

If, at the time of the `wait` call, the job is in the `NEW` state, which the example expects, `wait` will return when the job reaches the  `QUEUED` state.
If, at the time of the `wait` call, the job is already `ACTIVE`, an `UnreachableStateException` will eventually be thrown.

Whether it's one of the other depends on a number of things, such as: 
 - if and how quickly the underlying executor moves jobs through a queue
 - what exactly is in-between the `submit()` and `wait()` calls.

For reference, with a local executor which sets the job to `ACTIVE` during `submit()`, the `UnreacheableStateException` will always be triggered.

### Suggested Solution

One could explicitly wait for all states after `QUEUED`, but that just seems tedious and contrary to the way we intended this to work, which is to not have such silly complexities.

**Comment by andre-merzky** [comment]
I fully agree that the race exists - but that is not really avoidable in any system with concurrent threads of execution **when using a synchronous API** -- at least not with reasonable effort and performance.  I would argue that is is the responsibility of the user, and the proper solution is to use notifications:
```py
queued_signal = threading.Event()
def cb(job):
   if job.status.state == jpsi.QUEUED:
       queued_signal.set()
...
executor.submit(job)
...
queued_signal.wait()
...
```

**Comment by hategan** [comment]
I agree with you.

And then, maybe I don't, in the sense that I think that if we can eliminate race conditions by careful API design, we could do so. 

So the problem here is not as much that a race condition exists, but that it is largely connector dependent. To me, that means a broken abstraction, which is contrary to the purpose of the API. If this occurred in equal measure for all connectors, I'd be more inclined to dismiss it as unavoidable.

In the spirit of seeing of this is avoidable, it strikes me that when I say `wait([JobState.QUEUED])`, I mean "don't return before the job is queued". In other words, I don't read that as waiting for a particular transition, that of `JobState.NEW -> JobState.QUEUED`, but more as a `ensure_that_the_job_has_gone_through(JobState.QUEUED)`. 

Point being, if we allowed wait(x) to return for any state >= x, we wouldn't quite have this problem.

**Comment by andre-merzky** [comment]
> So the problem here is not as much that a race condition exists, but that it is largely connector dependent.

It just as much depends on when the user issues the second call which is not controllable and not avoidable.


> Point being, if we allowed wait(x) to return for any state >= x, we wouldn't quite have this problem.

Then `wait()` would never raise, am I reading this right?

**Comment by hategan** [comment]
> > So the problem here is not as much that a race condition exists, but that it is largely connector dependent.
> 
> It just as much depends on when the user issues the second call which is not controllable and not avoidable.

If that was the only thing it depends on, I'd be fine. But it's not. In particular, one of the examples in the spec fails here, so we didn't see this coming.

> 
> > Point being, if we allowed wait(x) to return for any state >= x, we wouldn't quite have this problem.
> 
> Then `wait()` would never raise, am I reading this right?

I don't think that's necessarily right. The easy one being waiting for COMPLETED and the job is canceled. More tricky, a job could jump from queued to failed, so waiting for ACTIVE would raise. Which is problematic when you consider missed events. But yeah, there's subtle issues here that might require a bit deeper analysis than I would have both expected and wanted.

**Comment by jameshcorbett** [comment]
What about if each `Job` kept track of all the events it has seen? Then `wait(jobstate.QUEUED)` would always return and never throw an `UnreachableStateException`, unless something happened to prevent the job from being queued. That is kind of how the [FluxExecutor in flux proper](https://github.com/flux-framework/flux-core/blob/master/src/bindings/python/flux/job/executor.py) works. You can attach an event callback to a `FluxExecutorFuture` at any point, and you won't miss any events. And some Flux events happen multiple times---that is handled too.

**Comment by hategan** [comment]
In principle, keeping track of all seen states can work, but only if the state transition diagram has no loops. Picture a scenario in which a job can go QUEUED -> ACTIVE -> PREEMPTED -> QUEUED -> ACTIVE, etc. Say you wait for PREEMPTED and then you wait for ACTIVE. That would return immediately since there was an ACTIVE that already happened.

So I'm with this, but I'd like to explore a solution that would work with arbitrary state diagrams, just in case things will evolve in that direction.

That said, there may not be a solution to this problem, in the sense that `submit(); wait()` or `wait(S1); wait(S2)` are fundamentally race-prone as opposed to adding a status callback before calling `submit()`. But I'd like to, if possible, have an "as-good-as-possible" synchronous solution, since I suspect many new users will play with synchronous code first.

**Comment by jameshcorbett** [comment]
We could offer two functions/methods, one which is racy and one which is not? If the functions are well-documented I don't think their shortcomings would be too problematic.

**Comment by jameshcorbett** [comment]
Or what about saving a thread-safe list of seen states and letting users access it? Then users could do whatever kinds of complicated logic they want to handle cycles and so on.

**Comment by hategan** [comment]
> Or what about saving a thread-safe list of seen states and letting users access it? Then users could do whatever kinds of complicated logic they want to handle cycles and so on.

I do like that. This can be implemented easily in userspace using the callback though, so maybe it shouldn't be in the job spec. The reason why I'd avoid it is that it storing all events for all jobs could be a big penalty hit that would apply whether that feature is used or not. So perhaps it's best left to user code to decide based on the trade-offs.

So maybe what Andre said initially makes the most sense, seeing that @jameshcorbett is roughly going in the same direction. If you want to keep track of all events, add a callback before submitting the job. If you want to use the sync API (aka. wait()), be careful. For example, if you want to wait for QUEUED, be mindful that the job might have already gone through that state and plan for that:

```python
status = job.wait([JobState.QUEUED, JobState.ACTIVE, JobState.COMPLETED])
if status.state == JobState.QUEUED:
    ...
elif...
```

So maybe the question is: are there any downsides to having `wait(QUEUED)` imply `wait(QUEUED, ACTIVE, COMPLETED)` and should we do that, since it is very likely that users will always have to write `wait(QUEUED, ACTIVE, COMPLETED)` otherwise?

**Comment by hategan** [comment]
The issue was discussed during our weekly meeting and the consensus appeared to be that (and please correct me if I'm wrong or missing something):
- `wait(s)` should imply wait(t)` for all t >= s; basically, it is unlikely for the current semantics of `wait(QUEUED)` to be what the user wants and it would be unnecessarily tedious to have to write `wait([QUEUED, ACTIVE, COMPLETED])`
- implicit states in `wait()` also make code more robust to changes in the state diagram; if a state 'STAGE_IN' is added between `QUEUED` and `ACTIVE`, without implicit states, all code would need to be updated to read `wait([QUEUED, STAGE_IN, ACTIVE, COMPLETED])`, whereas with implicit states no change would be necessary.
- since `wait()` returns the state that caused it to return, users can tell whether the desired state was reached or it was inferred.
- assuming that the job eventually terminates, `wait(...)` can either succeed or throw an `UnreachableStateException`, but never hang; this is not a consequence of implicit states, but a general statement about how we want `wait()` to behave
- should the state diagram be updated with loops, we can find ways to carefully allow states within a loop to be waited for without invalidating existing code; for example, if jobs can be kicked back to the queue and we want to reflect this in the states, we can add a `REQUEUED` state which is strictly greater than `QUEUED` but not comparable with anything after `QUEUED`; waiting for `REQUEUED` would then be strict.
- this was not discussed in the meeting: if the user wants the strict waiting behavior, `wait()` can be wrapped in a function that checks the returned state and throws `UnreachableStateException` as needed.

**Comment by andre-merzky** [comment]
> * this was not discussed in the meeting: if the user wants the strict waiting behavior, `wait()` can be wrapped in a function that checks the returned state and throws `UnreachableStateException` as needed.

True.  A `strict: bool = False` option to the wait call could offer that also.

**Comment by hategan** [comment]
> > * this was not discussed in the meeting: if the user wants the strict waiting behavior, `wait()` can be wrapped in a function that checks the returned state and throws `UnreachableStateException` as needed.
> 
> True. A `strict: bool = False` option to the wait call could offer that also.

I'd be in favor of keeping the spec lighter, especially since it's fairly easy to implement the strict behavior.
